### PR TITLE
Roles in conversations

### PR DESCRIPTION
**Pull request by tiago-loureiro**
Tasks: 
 
 - [x] Define types for conversation roles
 - [x] Define actions for conversation roles
 - [x] Define serialisation for conversation roles (JSON vs. bitmask)
 - [x] Define API for fetching conversation roles
 - [x] Add conversation role to conversation members
 - [x] Allow `wire_admin` to change the role of other members
 - [x] Allow settings member role during conversation creation
 - [x] Make `wire_admin` the default role in a conversation if none exists
 - [x] Make _actual_ use of conversation roles
 - [x] Ensure that custom roles cannot be created yet
 - [x] More integration tests

### CODE REVIEW

File: libs/galley-types/src/Galley/Types/Bot/Service/Internal.hs [hunk 1]
```diff
@@ -41,7 +44,11 @@ instance ToJSON ServiceRef where
 -- | A /secret/ bearer token used to authenticate and authorise requests @towards@
 -- a 'Service' via inclusion in the HTTP 'Authorization' header.
 newtype ServiceToken = ServiceToken AsciiBase64Url
-    deriving (Eq, Show, ToByteString, Cql, FromByteString, FromJSON, ToJSON, Generic)
+    deriving (Eq, Show, ToByteString, FromByteString, FromJSON, ToJSON, Generic)
+
+#ifdef WITH_CQL
+deriving instance Cql ServiceToken
+#endif
```
Comment (tiago-loureiro) [inline libs/galley-types/src/Galley/Types/Bot/Service/Internal.hs]: This change was done so that you can load ghci without the `cql` flag

File: services/galley/Makefile [hunk 2]
```diff
@@ -41,6 +41,10 @@ install: init
 fast: init
 	stack install . --pedantic --test --bench --no-run-benchmarks --local-bin-path=dist --fast $(WIRE_STACK_OPTIONS)
 
+.PHONY: ffast
```
Comment (tiago-loureiro) [inline services/galley/Makefile]: Ignore this, I will make a different PR for this and remove

File: services/galley/src/Galley/API/Update.hs [hunk 3]
```diff
@@ -333,73 +326,71 @@ joinConversation zusr zcon cnv access = do
...
     ensureMemberLimit (toList $ Data.convMembers conv) newUsers
     ensureAccess conv InviteAccess
     case Data.convTeam conv of
         Nothing -> do
-            ensureConvMember (snd mems) zusr
             ensureAccessRole (Data.convAccessRole conv) newUsers Nothing
             ensureConnected zusr newUsers
         Just ti -> teamConvChecks ti newUsers conv
-    addToConversation mems zusr zcon newUsers conv
+    addToConversation mems zusr zcon newUsers conv (invRoleName body)
```
Comment (jschaul) [inline services/galley/src/Galley/API/Update.hs]: This is insufficient: invRoleName should not be blindly trusted: 

* If this is a non-team conversation, ignore this and hardcode to wireAdmin. (don't kill democracy! :smile: )
* If this is a team conversation, first check this role exists, and is valid, and doesn't contain higher permission sets than the user granting, before proceeding
* If this is unset, set the default (wireAdmin for non-team conversations, wireMember for team conversations?), clearly explained with a comment.

File: services/galley/src/Galley/API/Update.hs [hunk 4]
```diff
@@ -333,73 +326,71 @@ joinConversation zusr zcon cnv access = do
...
-    applyChanges m u = m
-        { memOtrMuted       = fromMaybe (memOtrMuted m) (misOtrMuted u)
-        , memOtrMutedRef    = misOtrMutedRef u <|> memOtrMutedRef m
-        , memOtrArchived    = fromMaybe (memOtrArchived m) (misOtrArchived u)
-        , memOtrArchivedRef = misOtrArchivedRef u <|> memOtrArchivedRef m
-        , memHidden         = fromMaybe (memHidden m) (misHidden u)
-        , memHiddenRef      = misHiddenRef u <|> memHiddenRef m
-        }
+    toMemberUpdate :: OtherMemberUpdate -> MemberUpdate
+    toMemberUpdate omu = MemberUpdate Nothing Nothing Nothing Nothing Nothing Nothing Nothing (omuConvRoleName omu)
```
Comment (jschaul) [inline services/galley/src/Galley/API/Update.hs]: This is not pretty, can you use named fields here?

File: services/galley/src/Galley/Data.hs [hunk 5]
```diff
@@ -188,6 +190,16 @@ teamMembers t = mapM newTeamMember' =<<
     newTeamMember' (uid, perms, minvu, minvt, mlhStatus) =
         newTeamMemberRaw uid perms minvu minvt (fromMaybe UserLegalHoldDisabled mlhStatus)
 
+teamMembers' :: forall m. (MonadThrow m, MonadClient m) => TeamId -> [UserId] -> m [TeamMember]
```
Comment (jschaul) [inline services/galley/src/Galley/Data.hs]: What's the difference between `teamMembers'` and `teamMembers`? Shouldn't one use the other? Maybe add a comment here.

File: services/galley/src/Galley/Data.hs [hunk 6]
```diff
@@ -381,8 +393,9 @@ createConversation :: UserId
                    -> Maybe ConvTeamInfo
                    -> Maybe Milliseconds                  -- ^ Message timer
                    -> Maybe ReceiptMode
+                   -> RoleName
                    -> Galley Conversation
-createConversation usr name acc role others tinfo mtimer recpt = do
+createConversation usr name acc role others tinfo mtimer recpt othersRole = do
```
Comment (jschaul) [inline services/galley/src/Galley/Data.hs]: `othersRole`? `conversationRole` ?

File: services/galley/src/Galley/Data/Queries.hs [hunk 7]
```diff
@@ -239,8 +255,8 @@ selectSrv = "select base_url, auth_token, fingerprints, enabled from service whe
 
 -- Bots ---------------------------------------------------------------------
 
-insertBot :: PrepQuery W (ConvId, BotId, ServiceId, ProviderId) ()
-insertBot = "insert into member (conv, user, service, provider, status) values (?, ?, ?, ?, 0)"
+insertBot :: PrepQuery W (ConvId, BotId, ServiceId, ProviderId, RoleName) ()
+insertBot = "insert into member (conv, user, service, provider, status, conversation_role) values (?, ?, ?, ?, 0, ?)"
```
Comment (jschaul) [inline services/galley/src/Galley/Data/Queries.hs]: Please document how/why bots need a conversation_role. Don't they already not have access to that part of the API? At the moment they're whitelisted to only make requests to `/bot/` endpoints. Maybe revert this change?

File: libs/galley-types/src/Galley/Types.hs [hunk 8]
```diff
@@ -857,81 +898,121 @@ instance FromJSON MemberUpdate where
...
+
+instance ToJSON OtherMemberUpdate where
+    toJSON m = object
+        $ "conversation_role" .= omuConvRoleName m
         # []
 
 instance FromJSON MemberUpdateData where
     parseJSON = withObject "member-update event data" $ \m ->
-        MemberUpdateData <$> m .:? "otr_muted"
+        MemberUpdateData <$> m .:? "id"
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types.hs]: >  already has an evtFrom :: !UserId

Correct, that is a `From`, i.e., who performed the action. You don't know which user this action was perform _on_, which is why this was introduced versus introducing a totally separate event (I think this is a nice improvement and not just avoiding introducing a new type)

File: libs/galley-types/src/Galley/Types.hs [hunk 9]
```diff
@@ -946,12 +1027,31 @@ instance ToJSON ConvType where
...
-instance ToJSON Members where
-    toJSON e = object [ "user_ids" .= mUsers e]
+instance FromJSON UserIdList where
+    parseJSON = withObject "user-ids-payload" $ \o ->
+        UserIdList <$> o .: "user_ids"
+
+instance ToJSON UserIdList where
+    toJSON e = object [ "user_ids" .= mUsers e ]
+
+-- TODO: Think about backwards compatibility here...
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types.hs]: Not relevant anymore, I will remove it since it was taken care of during the coding phase.

File: libs/galley-types/src/Galley/Types/Bot.hs [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE OverloadedStrings          #-}
 {-# LANGUAGE TemplateHaskell            #-}
+-- {-# LANGUAGE StandaloneDeriving         #-}
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Bot.hs]: Will be removed, thanks!

File: libs/galley-types/src/Galley/Types/Conversations/Roles.hs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+--  * Custom RoleNames _must not_ start with `wire_`
+-- TODO: _Actually_ ensure all the above properties
+isValidRoleName :: Text -> Bool
+isValidRoleName = either (const False) (const True)
+                . parseOnly customRoleName
+  where
+    customRoleName = count 2 (satisfy chars)
+                  *> count 126 (optional (satisfy chars))
+                  *> endOfInput
+    chars = inClass "a-zA-Z0-9_-"
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Conversations/Roles.hs]: Good idea 👍

**Comment by tiago-loureiro** [comment]
> There is big backwards compatibility issue with respect to the permission model: All existing conversations in a team, where at the moment some people do not have the permission to add members to a conversation, will receive conversation admin rights, since all the team user checks are abolished and existing conversation members are wire_admin

Right, this is known and by design

File: services/galley/src/Galley/API/Teams.hs [hunk 12]
```diff
@@ -410,16 +422,17 @@ getTeamConversation (zusr::: tid ::: cid ::: _) = do
 
 deleteTeamConversation :: UserId ::: ConnId ::: TeamId ::: ConvId ::: JSON -> Galley Response
 deleteTeamConversation (zusr::: zcon ::: tid ::: cid ::: _) = do
-    tmems <- Data.teamMembers tid
-    void $ permissionCheck zusr DeleteConversation tmems
+    -- FUTUREWORK: Previously, any team member could delete the
+    --             conversation which is a weird feature? In any
+    --             case, do not lookup all team members here
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Teams.hs]: Actually I fixed that, ignore this.

File: services/galley/src/Galley/API/Util.hs [hunk 13]
```diff
@@ -162,3 +174,33 @@ convMembsAndTeamMembs convMembs teamMembs
 membersToRecipients :: Maybe UserId -> [TeamMember] -> [Recipient]
 membersToRecipients Nothing  = map (userRecipient . view userId)
 membersToRecipients (Just u) = map userRecipient . filter (/= u) . map (view userId)
+
+-- Note that we use 2 nearly identical functions but slightly different
+-- semantics; when using `getSelfMember`, if that user is _not_ part of
+-- the conversation, we don't want to disclose that such a conversation
+-- with that id exists.
+getSelfMember :: Foldable m => UserId -> m Member -> Galley Member
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Util.hs]: > This is a foldable/traversable (List) of Member, not a monad, right?

Nice one!

File: libs/galley-types/src/Galley/Types/Teams.hs [hunk 14]
```diff
@@ -242,11 +242,11 @@ data Permissions = Permissions
...
 data Perm =
       CreateConversation
-    | DeleteConversation
+    | DeleteConversation  -- NOTE: This gets now overruled by conv level checks
     | AddTeamMember
     | RemoveTeamMember
-    | AddRemoveConvMember
-    | ModifyConvMetadata
+    | AddRemoveConvMember -- NOTE: This gets now overruled by conv level checks
+    | ModifyConvName      -- NOTE: This gets now overruled by conv level checks
```
Comment (jschaul) [inline libs/galley-types/src/Galley/Types/Teams.hs]: How about you rename all of these to `NotInUseDeprecatedXX` or remove them entirely?

File: libs/galley-types/src/Galley/Types.hs [hunk 15]
```diff
@@ -337,15 +351,30 @@ data MemberUpdate = MemberUpdate
...
     , mupHiddenRef     :: !(Maybe Text)
+    , mupConvRoleName  :: !(Maybe RoleName)
     }
 
 deriving instance Eq   MemberUpdate
 deriving instance Show MemberUpdate
 
-newtype Invite = Invite
-    { invUsers :: List1 UserId
+-- Inbound other member updates.  This is what galley expects on its endpoint.  See also
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types.hs]: Done

File: libs/galley-types/src/Galley/Types/Swagger.hs [hunk 16]
```diff
@@ -386,10 +415,23 @@ memberUpdate = defineModel "MemberUpdate" $ do
...
+    description "Update user properties of other members relative to a conversation"
+    property "conversation_role" string' $ do
+        description "Name of the conversation role updated to"
+        optional
 
 memberUpdateData :: Model
 memberUpdateData = defineModel "MemberUpdateData" $ do
     description "Event data on member updates"
+    property "id" bytes' $ do
+        description "Whether to notify on conversation updates"
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Swagger.hs]: Thanks!

File: services/brig/src/Brig/Provider/API.hs [hunk 17]
```diff
@@ -727,7 +728,7 @@ addBot (zuid ::: zcon ::: cid ::: req) = do
     let bcl = newClientId (fromIntegral (hash bid))
 
     -- Ask the external service to create a bot
-    let origmem = OtherMember zuid Nothing
+    let origmem = OtherMember zuid Nothing roleNameWireMember
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/brig/src/Brig/Provider/API.hs]: 👍

File: services/galley/src/Galley/API/Util.hs [hunk 18]
```diff
@@ -61,6 +63,16 @@ ensureReAuthorised u secret = do
...
 
+-- | Given a member in a conversation, check if the given action
+-- is permitted.
+-- If not, throw 'Member'; if the user is found and does not have the given permission, throw
+-- 'operationDenied'.  Otherwise, return the found user.
+ensureActionAllowed :: Action -> Member -> Galley ()
+ensureActionAllowed action mem = case isActionAllowed action (memConvRoleName mem) of
+    Just True  -> return ()
+    Just False -> throwM (actionDenied action)
+    Nothing    -> throwM notImplemented
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Util.hs]: Sending a 400 instead. Note that the code will be so that no such codepath can exist (although not ensured by types)

File: services/galley/src/Galley/Data.hs (hunk 5 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/Data.hs]: I renamed `teamMembers'` to `teamMembersLimited` with a comment, hope that's more clear. One cannot reuse the other since the CQL query needs to look quite different.

File: libs/galley-types/src/Galley/Types.hs [hunk 19]
```diff
@@ -409,15 +443,17 @@ data Connect = Connect
...
+-- 'MemberUpdate' and 'OtherMemberUpdate'.
 data MemberUpdateData = MemberUpdateData
-    { misOtrMuted       :: !(Maybe Bool)
-    , misOtrMutedStatus :: !(Maybe MutedStatus)
-    , misOtrMutedRef    :: !(Maybe Text)
-    , misOtrArchived    :: !(Maybe Bool)
-    , misOtrArchivedRef :: !(Maybe Text)
-    , misHidden         :: !(Maybe Bool)
-    , misHiddenRef      :: !(Maybe Text)
+    { misId               :: !(Maybe UserId)
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types.hs]: Yes, renamed to `target`

File: services/galley/src/Galley/API.hs [hunk 20]
```diff
@@ -678,6 +715,27 @@ sitemap = do
...
+        .&. capture "usr"
+        .&. jsonRequest @OtherMemberUpdate
+
+    document "PUT" "updateOtherMember" $ do
+        summary "Update membership of the specified user"
+        notes "Even though all fields are optional, at least one needs to be given."
+        parameter Path "cnv" bytes' $
+            description "Conversation ID"
+        parameter Path "usr" bytes' $
+            description "User ID"
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API.hs]: I changed this in the event as I agree that it makes things more clear; however, at this level, I am tempted to change only the comment (done).

Reason for that being that we use `:usr` (or `:uid`) on other endpoints (e.g. `delete "/conversations/:cnv/members/:usr"`). Do u think that changing this to `target` makes the endpoint more readable?

File: libs/galley-types/src/Galley/Types/Swagger.hs [hunk 21]
```diff
@@ -386,10 +415,23 @@ memberUpdate = defineModel "MemberUpdate" $ do
...
+otherMemberUpdate = defineModel "otherMemberUpdate" $ do
+    description "Update user properties of other members relative to a conversation"
+    property "conversation_role" string' $ do
+        description "Name of the conversation role updated to"
+        optional
 
 memberUpdateData :: Model
 memberUpdateData = defineModel "MemberUpdateData" $ do
     description "Event data on member updates"
+    property "id" bytes' $ do
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Swagger.hs]: Done

File: libs/galley-types/src/Galley/Types/Swagger.hs [hunk 22]
```diff
@@ -408,6 +450,18 @@ memberUpdateData = defineModel "MemberUpdateData" $ do
...
         description "A reference point for (un)hiding"
         optional
+    property "conversation_role" string' $ do
+        description "Name of the conversation role to update to"
+        optional
+
+otherMemberUpdateData :: Model
+otherMemberUpdateData = defineModel "OtherMemberUpdateData" $ do
+    description "Event data on other member updates"
+    property "id" bytes' $
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Swagger.hs]: Done

File: services/galley/src/Galley/API/Create.hs [hunk 23]
```diff
@@ -153,7 +154,7 @@ createConnectConversation (usr ::: conn ::: req) = do
             | usr `isMember` mems -> connect n j conv
             | otherwise           -> do
                 now <- liftIO getCurrentTime
-                mm  <- snd <$> Data.addMember now (Data.convId conv) usr
+                mm  <- snd <$> Data.addMember now (Data.convId conv) usr roleNameWireAdmin
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Create.hs]: Yes, I removed that, 👍

File: services/galley/src/Galley/API/Teams.hs [hunk 24]
```diff
@@ -494,7 +507,7 @@ addTeamMemberInternal tid origin originConn newMem mems = do
     cc  <- filter (view managedConversation) <$> Data.teamConversations tid
     now <- liftIO getCurrentTime
     for_ cc $ \c ->
-        Data.addMember now (c^.conversationId) (new^.userId)
+        Data.addMember now (c^.conversationId) (new^.userId) roleNameWireAdmin
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Teams.hs]: I removed this change as well.

File: services/galley/src/Galley/API/Update.hs (hunk 4 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Update.hs]: Yes, I improved that 👍

File: services/galley/src/Galley/Data/Queries.hs (hunk 7 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/Data/Queries.hs]: Agreed, removed references to any roles when bots are concerned.

File: libs/galley-types/src/Galley/Types/Teams.hs (hunk 14 above)
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Teams.hs]: I renamed them because removing them will break clients.

File: libs/galley-types/src/Galley/Types/Swagger.hs [hunk 25]
```diff
@@ -102,10 +104,37 @@ connectEvent = defineModel "ConnectEvent" $ do
...
+conversationNameUpdateEvent = defineModel "ConversationNameUpdateEvent" $ do
     description "conversation update event"
-    property "data" (ref conversationUpdate) $ description "conversation data"
+    property "data" (ref conversationUpdateName) $ description "conversation name"
+
+conversationRole :: Model
+conversationRole = defineModel "ConversationRole" $ do
+    description "Conversation role"
+    property "conversation_role" string' $
+        description "role name, between 2 and 128 chars"
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Swagger.hs]: I made a slightly different comment

File: services/galley/src/Galley/API.hs [hunk 26]
```diff
@@ -448,17 +479,23 @@ sitemap = do
 
     ---
 
-    put "/conversations/:cnv" (continue updateConversation) $
+    -- FUTUREWORK:
+    --       We should deprecate this in favor of
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/galley/src/Galley/API.hs]: Good idea, added. I will ping clients teams afterwards.

File: libs/galley-types/src/Galley/Types/Conversations/Roles.hs [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+
+parseRoleName :: Text -> Maybe RoleName
+parseRoleName t
+    | isValidRoleName t = Just (RoleName t)
+    | otherwise         = Nothing
+
+-- All RoleNames should have 2-128 chars
+--  * Wire RoleNames _must_ start with `wire_`
+--  * Custom RoleNames _must not_ start with `wire_`
+-- TODO: _Actually_ ensure all the above properties
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/galley-types/src/Galley/Types/Conversations/Roles.hs]: This has now been fixed too https://github.com/wireapp/wire-server/commit/2448362a5e8c9e9d33c42d641bdc0e2a2ec413b3

File: services/galley/src/Galley/Data.hs (hunk 6 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/Data.hs]: Right, I renamed it to try and be more explicit; it refers to the role that `others` should have: 616ff0edeb353a1018efd268a5fe0d7c07f468ca

File: services/brig/test/integration/API/Team/Util.hs [hunk 28]
```diff
@@ -101,7 +102,7 @@ createTeamConv :: HasCallStack => Galley -> TeamId -> UserId -> [UserId] -> Mayb
 createTeamConv g tid u us mtimer = do
     let tinfo = Just $ ConvTeamInfo tid False
     let conv = NewConvUnmanaged $
-               NewConv us Nothing (Set.fromList []) Nothing tinfo mtimer Nothing
+               NewConv us Nothing (Set.fromList []) Nothing tinfo mtimer Nothing roleNameWireAdmin
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/brig/test/integration/API/Team/Util.hs]: Check the test named `createTeamConvLegacy` for this.

File: services/galley/src/Galley/API/Update.hs (hunk 3 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/API/Update.hs]: The check for elevation was added with `ensureConvRoleNotElevated`, thanks for pointing that out!
